File: src/ns/envelope.ts
import { Ableton } from "../index.js";
import { Namespace } from "./index.js";
import type { RawClip } from "./clip.js";
import { DeviceParameter, RawDeviceParameter } from "./device-parameter.js";

export interface EnvelopeEventControlCoefficients {
  x1: number;
  y1: number;
  x2: number;
  y2: number;
}

export interface EnvelopeEvent {
  time: number;
  value: number;
  control_coefficients: EnvelopeEventControlCoefficients;
}

export interface GettableProperties {
  parameter: RawDeviceParameter;
}

export interface TransformedProperties {
  parameter: DeviceParameter;
}

export interface SettableProperties {}

export interface ObservableProperties {}

export interface RawEnvelope {
  readonly id: string;
}

/**
 * An automation or modulation envelope in Live.
 */
export class Envelope extends Namespace<
  GettableProperties,
  TransformedProperties,
  SettableProperties,
  ObservableProperties
> {
  constructor(
    ableton: Ableton,
    public readonly raw: RawEnvelope,
  ) {
    super(ableton, "envelope", raw.id);

    this.transformers = {
      parameter: (param) => new DeviceParameter(ableton, param),
    };

    this.cachedProps = {
      parameter: true,
    };
  }

  /**
   * Creates a new event at the specified time with the given value and,
   * optionally, control coefficients.
   */
  createEvent(
    time: number,
    value: number,
    controlCoefficients?: EnvelopeEventControlCoefficients,
  ): Promise<void> {
    return this.sendCommand("create_event", {
      time,
      value,
      control_coefficients: controlCoefficients,
    });
  }

  /**
   * Deletes the events in the specified time range.
   */
  deleteEventsInRange(fromTime: number, toTime: number): Promise<void> {
    return this.sendCommand("delete_events_in_range", {
      from_time: fromTime,
      to_time: toTime,
    });
  }

  /**
   * Returns the events in the specified time range.
   */
  eventsInRange(fromTime: number, toTime: number): Promise<EnvelopeEvent[]> {
    return this.sendCommand("events_in_range", {
      from_time: fromTime,
      to_time: toTime,
    });
  }

  /**
   * Given a start time, a step length and a value, creates a step in the envelope.
   */
  insertStep(startTime: number, length: number, value: number): Promise<void> {
    return this.sendCommand("insert_step", {
      start_time: startTime,
      length,
      value,
    });
  }

  /**
   * Returns the parameter value at the specified time.
   */
  valueAtTime(time: number): Promise<number> {
    return this.sendCommand("value_at_time", { time });
  }
}

File: midi-script/Namespaces/Envelope.py

import Live

from .DeviceParameter import DeviceParameter
from .Interface import Interface


class Envelope(Interface):
    @staticmethod
    def serialize_envelope(envelope):
        if envelope is None:
            return None

        envelope_id = Interface.save_obj(envelope)
        return {"id": envelope_id}

    @staticmethod
    def serialize_event(event):
        if event is None:
            return None

        coefficients = event.control_coefficients
        return {
            "time": event.time,
            "value": event.value,
            "control_coefficients": {
                "x1": coefficients.x1,
                "y1": coefficients.y1,
                "x2": coefficients.x2,
                "y2": coefficients.y2,
            },
        }

    def __init__(self, c_instance, socket):
        super().__init__(c_instance, socket)

    def get_parameter(self, ns):
        return DeviceParameter.serialize_device_parameter(ns.parameter)

    def create_event(self, ns, time, value, control_coefficients=None):
        event_time = float(time)
        event_value = float(value)
        if control_coefficients is None:
            event = Live.Envelope.EnvelopeEvent(event_time, event_value)
        else:
            coefficients = Live.Envelope.EnvelopeEventControlCoefficients(
                float(control_coefficients.get("x1", 0.5)),
                float(control_coefficients.get("y1", 0.5)),
                float(control_coefficients.get("x2", 0.5)),
                float(control_coefficients.get("y2", 0.5)),
            )
            event = Live.Envelope.EnvelopeEvent(event_time, event_value, coefficients)
        ns.create_event(event)

    def delete_events_in_range(self, ns, from_time, to_time):
        ns.delete_events_in_range(float(from_time), float(to_time))

    def events_in_range(self, ns, from_time, to_time):
        events = ns.events_in_range(float(from_time), float(to_time))
        return [Envelope.serialize_event(event) for event in events]

    def insert_step(self, ns, start_time, length, value):
        ns.insert_step(float(start_time), float(length), float(value))

    def value_at_time(self, ns, time):
        return ns.value_at_time(float(time))
